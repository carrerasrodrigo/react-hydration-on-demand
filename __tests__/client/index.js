/**
 * @jest-environment jsdom
 */

import React from "react";
import ReactDom from "react-dom";
import { render, fireEvent, wait } from "@testing-library/react";

import withHydrationOnDemand from "../../src";

const Component = ({ label }) => <div className="label">{label}</div>;
const serverSideText = "some content server side";
const clientSideText = "some content client side";

const SSRhtml = `<section data-no-hydrate="true"><div class="label">${serverSideText}</div></section>`;

describe("With SSR", () => {
  test("Render correctly client side, no option ", () => {
    const elem = document.createElement("div");
    elem.innerHTML = SSRhtml;

    const ComponentWithHydrationOnDemandClient = withHydrationOnDemand()(
      Component
    );

    render(<ComponentWithHydrationOnDemandClient label={clientSideText} />, {
      container: elem,
      hydrate: true
    });

    expect(elem).toMatchSnapshot();
  });

  test("Render correctly client side, with delay", async () => {
    const elem = document.createElement("div");
    elem.innerHTML = SSRhtml;

    const ComponentWithHydrationOnDemandClient = withHydrationOnDemand({
      on: [["delay", 1000]]
    })(Component);

    const { getByText } = render(
      <ComponentWithHydrationOnDemandClient label={clientSideText} />,
      {
        container: elem,
        hydrate: true
      }
    );

    await wait(() => getByText(clientSideText));

    expect(elem).toMatchSnapshot();
  });

  test("Render correctly client side, with DOM event (onClick)", async () => {
    const elem = document.createElement("div");
    elem.innerHTML = SSRhtml;

    const ComponentWithHydrationOnDemandClient = withHydrationOnDemand({
      on: ["click"]
    })(Component);

    const { getByText } = render(
      <ComponentWithHydrationOnDemandClient label={clientSideText} />,
      {
        container: elem,
        hydrate: true
      }
    );

    fireEvent.click(getByText(serverSideText));

    await wait(() => getByText(clientSideText));

    expect(elem).toMatchSnapshot();
  });

  test("Render correctly client side, with DOM event (onClick), custom selector", async () => {
    const elem = document.createElement("div");
    elem.innerHTML = SSRhtml;

    const ComponentWithHydrationOnDemandClient = withHydrationOnDemand({
      on: [["click", elem]]
    })(Component);

    const { getByText } = render(
      <ComponentWithHydrationOnDemandClient label={clientSideText} />,
      {
        container: elem,
        hydrate: true
      }
    );

    fireEvent.click(elem);

    await wait(() => getByText(clientSideText));

    expect(elem).toMatchSnapshot();
  });

  test("Render correctly client side, on idle", async () => {
    const elem = document.createElement("div");
    elem.innerHTML = SSRhtml;

    window.requestIdleCallback = jest.fn(f => f());

    const ComponentWithHydrationOnDemandClient = withHydrationOnDemand({
      on: ["idle"]
    })(Component);

    const { getByText } = render(
      <ComponentWithHydrationOnDemandClient label={clientSideText} />,
      {
        container: elem,
        hydrate: true
      }
    );

    await wait(() => getByText(clientSideText));

    expect(window.requestIdleCallback).toHaveBeenCalled();
    expect(elem).toMatchSnapshot();
  });

  test("Client side, execute onBefore before hydration", async () => {
    const elem = document.createElement("div");
    elem.innerHTML = SSRhtml;

    const onBefore = jest.fn();

    const ComponentWithHydrationOnDemandClient = withHydrationOnDemand({
      on: ["click"],
      onBefore
    })(Component);

    const { getByText } = render(
      <ComponentWithHydrationOnDemandClient label={clientSideText} />,
      {
        container: elem,
        hydrate: true
      }
    );

    fireEvent.click(getByText(serverSideText));

    await wait(() => getByText(clientSideText));

    expect(onBefore).toHaveBeenCalled();
  });

  test("Render correctly client side, forceHydration ", async () => {
    const elem = document.createElement("div");
    elem.innerHTML = SSRhtml;

    const ComponentWithHydrationOnDemandClient = withHydrationOnDemand()(
      Component
    );

    const { getByText } = render(
      <ComponentWithHydrationOnDemandClient
        forceHydration
        label={clientSideText}
      />,
      {
        container: elem,
        hydrate: true
      }
    );

    await wait(() => getByText(clientSideText));

    expect(elem).toMatchSnapshot();
  });
});

describe("Without SSR", () => {
  test("Render correctly client side", async () => {
    const elem = document.createElement("div");

    const ComponentWithHydrationOnDemandClient = withHydrationOnDemand()(
      Component
    );

    const { getByText } = render(
      <ComponentWithHydrationOnDemandClient label={clientSideText} />,
      {
        container: elem,
        hydrate: true
      }
    );

    await wait(() => getByText(clientSideText));

    expect(elem).toMatchSnapshot();
  });
});
