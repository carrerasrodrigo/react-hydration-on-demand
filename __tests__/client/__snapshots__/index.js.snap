// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`With SSR Render correctly client side, forceHydration  1`] = `
<div>
  <section
    data-no-hydrate="true"
  >
    <div
      class="label"
    >
      some content client side
    </div>
  </section>
</div>
`;

exports[`With SSR Render correctly client side, no option  1`] = `
<div>
  <section
    data-no-hydrate="true"
  >
    <div
      class="label"
    >
      some content server side
    </div>
  </section>
</div>
`;

exports[`With SSR Render correctly client side, on idle 1`] = `
<div>
  <section
    data-no-hydrate="true"
  >
    <div
      class="label"
    >
      some content client side
    </div>
  </section>
</div>
`;

exports[`With SSR Render correctly client side, with DOM event (onClick) 1`] = `
<div>
  <section
    data-no-hydrate="true"
  >
    <div
      class="label"
    >
      some content client side
    </div>
  </section>
</div>
`;

exports[`With SSR Render correctly client side, with DOM event (onClick), custom selector 1`] = `
<div>
  <section
    data-no-hydrate="true"
  >
    <div
      class="label"
    >
      some content client side
    </div>
  </section>
</div>
`;

exports[`With SSR Render correctly client side, with delay 1`] = `
<div>
  <section
    data-no-hydrate="true"
  >
    <div
      class="label"
    >
      some content client side
    </div>
  </section>
</div>
`;

exports[`Without SSR Render correctly client side 1`] = `
<div>
  <section
    style="display: inline-block;"
  >
    <div
      class="label"
    >
      some content client side
    </div>
  </section>
</div>
`;
